lib/site-inspector.rb

```diff
@@ -127,37 +134,213 @@ def to_json
-      :advertising => advertising,
-      :slash_data => slash_data?,
-      :slash_developer => slash_developer?,
-      :data_dot_json => data_dot_json?,
-      :click_jacking_protection => click_jacking_protection?,
-      :content_security_policy => content_security_policy?,
-      :xss_protection => xss_protection?,
-      :secure_cookies => secure_cookies?,
-      :strict_transport_security => strict_transport_security?,
-      :headers => headers
+      https: {
+        www: https_www,
+        root: https_root
+      },
+      http: {
+        www: http_www,
+        root: http_root
+      }
     }
   end
+
+  # State of affairs at a particular endpoint.
+  def http_endpoint(ssl, www)
+    details = {}
+
+    # Don't follow redirects for first ping.
+    response = request(ssl, www, false)
+
+
+    # For HTTPS: examine the full range of possibilities.
+    if ssl
+      if response.return_code == :ok
+        details[:https_valid] = true
+
+      # Bad certificate chain.
+      elsif response.return_code == :ssl_cacert
+        details[:https_valid] = false
+        details[:https_bad_chain] = true
+        response = request(ssl, www, false, false, true)
+        # Bad everything.
+        if response.return_code == :peer_failed_verification
+          details[:https_bad_name] = true
+          response = request(ssl, www, false, false, false)
+        end
+      # Bad hostname.
+      elsif response.return_code == :peer_failed_verification
+        details[:https_valid] = false
+        details[:https_bad_name] = true
+        response = request(ssl, www, false, true, false)
+        # Bad everything.
+        if response.return_code == :ssl_cacert
+          details[:https_bad_chain] = true
+          response = request(ssl, www, false, false, false)
+        end
+      end
+    end
+
+    # If we ended up with a failure, return it.
+    details[:status] = response.response_code
+    return details if response.response_code == 0
+
+    headers = Hash[response.headers.map{ |k,v| [k.downcase,v] }]
+    # details[:headers] = headers
+
+
+    # HSTS only takes effect when delivered over valid HTTPS.
+    details[:hsts] = !!(ssl and details[:https_valid] and headers["strict-transport-security"])
+    details[:hsts_header] = headers["strict-transport-security"]
+
+
+    # If it's a redirect, go find the ultimate response starting from this combo.
+    redirect_code = response.response_code.to_s.start_with?("3")
+    location_header = headers["location"]
+    if redirect_code and location_header
+      details[:redirect] = true
+
+      ultimate_response = request(ssl, www, true, !details[:https_bad_chain], !details[:https_bad_name])
+      uri_original = URI(ultimate_response.request.url)
+      uri_eventual = URI(ultimate_response.effective_url)
+
+      details[:redirect_to] = uri_eventual.to_s
+      details[:redirect_to_external] = ((uri_original.hostname != uri_eventual.hostname) or (uri_original.scheme != uri_eventual.scheme))
```

Comment:
An "external" redirect, from the perspective of an individual endpoint, is one that redirects to any other endpoint. So a redirect from `http://whitehouse.gov` to `https://whitehouse.gov` is considered "external" here.

The whole domain is considered a "redirector" if all 4 domains have external redirects. The only situation I can imagine where the 4 endpoints all redirect amongst themselves without leaving to another domain is if there's a redirect loop -- so it's possible that a domain which is rendered totally inaccessible through redirect loops at all 4 endpoints would be incorrectly considered a "redirector". I'm okay with that (and have never observed such a domain).